### ISSUE TITLE
How to register custom CRS and CRS coordinate operations ?

### ISSUE THREAD

**Issue by jsorel** [proj4js/proj4js#488]
Hello,

I would like to handle custom CRS and it's related coordinate operations in Proj4js (and OpenLayer indirectly).

I have 2 questions :

1) Imagine you have a custom CRS (something proj4js can not handle, odd grids or not on earth), how do I create and register this CRS ?

2) For this CRS, I can provide the Javascript function to forward/inverse transform coordinates from this CRS to any other CRS. How can I register a service proj4js can query to have the transform function ?

Thank you

**Comment by ahocevar** [comment]
OpenLayers can do arbitrary transform functions, Proj4js cannot.

**Comment by jsorel** [comment]
Would you be okay if I provide a patch to add a simple CoordinateOperationFactory API (ISO 19111) ?

**Comment by ahocevar** [comment]
@jsorel To be honest, I'm not clear about what you want to achieve. Could you provide an API example of what you're envisioning?

**Comment by jsorel** [comment]
Here is the new file coordinateOperations.js :

```
/**
 * Simplified version of ISO 19111 : Geographic information — Referencing by coordinates
 * 
 * CoordinateOperation
 * @typedef {Object} CoordinateOperation
 * @property {function} transform - Transform a point, returns the transformed point
 * 
 * CoordinateOperationFactory
 * @typedef {Object} CoordinateOperationFactory
 * @property {function} createOperation - Create a CoordinateOperation from source to target projection, result is null if not supported
 */

/**
 * Store CoordinateOperationFactory.
 */
var coordinateOperationFactories = [];

/**
 * Register a CoordinateOperationFactory.
 *
 * @param {CoordinateOperationFactory} factory 
 */
export function registerCoordinateOperationFactory(factory) {
  if (factory !== null) {
    coordinateOperationFactories.push(factory);
  }
}

/**
 * Get coordinate operation from source to target.
 *
 * @param {*} source 
 * @param {*} dest 
 * @returns 
 */
export default function findCoordinateOperation(source, dest) {
  for (var cof in coordinateOperationFactories) {
    var co = cof.createOperation(source, dest);
    if (co !== null) {
      return co;
    }
  }
  return null;
}
```

and a small modification in transform.js adding line 29 :
```
  //to use coordinate operation factories, projections must be explicit
  if (source && dest) {
    var co = findCoordinateOperation(source, dest);
    if (co) {
      return co.transform(point);
    }
  }
```

**Comment by jsorel** [comment]
Here is an example of what returns our server  for CRS:84 to EPSG:3395  (I know proj4js can do this, but it's just an example).

```
/*
 * The valid geographic area for the given coordinate operation (as an array [west, south, east, north]), or undefined
 */
const areaOfValidity = [-180.0, -80.0, 180.0, 84.0];

/*
 * Positional accuracy estimation in metres for the given operation, or NaN if unknown.
 */
const accuracy = 0.0;

/*
 * The mathematical formula to transform coordinates
 */
function transform(src){

	// STEP 1 : AffineTransform2D
	dst = new Array(2);
	/*
	┌                                                ┐
	│  0                     0.017453292519943295  0 │
	│ -0.017453292519943295  0                     0 │
	│  0                     0                     1 │
	└                                                ┘
	*/
	dst[0] = 0.017453292519943295 * src[1];
	dst[1] = -0.017453292519943295 * src[0];
	src = dst;

	// STEP 2 : Mercator
	dst = new Array(2);
	const ANGULAR_TOLERANCE = 1.5706706731410455E-9;
	const eccentricity = 0.08181919084262244;
	const expΨ = (φ, ℯsinφ) => Math.tan(Math.PI/4 + 0.5*φ) * Math.pow((1 - ℯsinφ) / (1 + ℯsinφ), 0.5*eccentricity);
	const copySign = (x, y) => Math.sign(x) === Math.sign(y) ? x : -x;
	φ = src[1];
	sinφ = Math.sin(φ);
	if (φ == 0.0) {
	  y = φ;
	} else {
	  a = Math.abs(φ);
	  if (a < Math.PI / 2.0) {
	    y = Math.log(expΨ(φ, eccentricity * sinφ));
	  } else if (a <= (Math.PI/2 + ANGULAR_TOLERANCE)) {
	    y = copySign(Number.POSITIVE_INFINITY, φ);
	  } else {
	    y = NaN;
	  }
	}
	dst[0] = src[0];
	dst[1] = y;
	src = dst;

	// STEP 3 : AffineTransform2D
	dst = new Array(2);
	/*
	┌                          ┐
	│ 6378137.0         0    0 │
	│       0    -6378137.0  0 │
	│       0           0    1 │
	└                          ┘
	*/
	dst[0] = 6378137.0 * src[0];
	dst[1] = -6378137.0 * src[1];
	src = dst;

	return dst;
}

module.exports = { transform, inverseTransform, accuracy, areaOfValidity }
```

to use it in Proj4JS, I expect the code to look like : 

```
//load the above transform function
var myTransform = ...

var myFactory = {
  createOperation : function(source, target) {
     if (source.projName == 'CRS:84' && target.projName == 'EPSG:3395') {
       return {
         transform = function(pt) {
            var res = myTransform([pt.x, pt.y]);
            pt.x = res[0];
            pt.y = res[1];
            return pt;
         }
       };
     }
     return null;
  }
};


proj4.registerCoordinateOperationFactory(myFactory);

//call a transform coordinate as usual with proj4js, nothing changes.

```

**Comment by ahocevar** [comment]
Maybe I'm failing to understand what your aiming for, or you're not describing it well. But I don't see why you would need proj4js here:
```js
const src = [x, y]
const dst = transform(src);
console.log(dst); // [transformedX, transformedY]
```
is all you need with your function.

**Comment by ahocevar** [comment]
I see. In OpenLayers, you can use `addCoordinateTransforms()` for that. See https://openlayers.org/en/latest/examples/wms-custom-proj.html. Would that be sufficient? If not, I think we have to work on a more intuitive API for registering such transforms, maybe something along the lines of Proj's coordinate operations.

**Comment by jsorel** [comment]
For OpenLayer the addCoordinateTransforms will work I think yes (I'm already playing with this demo).

I don't know ProJ coordinate operation API, but since a few versions ago Proj has been refactored using Apache SIS as model and the above proposal is a light version of Apache SIS API. I believe and hope they won't be too different.


How do you imagine the API ?

**Comment by desruisseaux** [comment]
Hello @ahocevar. Thanks for your answer. Please allow me to put some context. We are working for a space agency which would like to be able to use OpenLayer, or similar web applications, on Moon, Mars and other astronomical bodies. Those web applications use Proj4JS for their map projections. But Proj4JS does not (as far as I know, I may be wrong) contains CRS definitions for planets other than Earth. Currently, space agencies cheat by pretending that a map of Mars is on Earth. This approach can somewhat okay for visualization with spherical formulas, but become problematic for more advanced applications, e.g. when taking in account the ellipsoid is desired (the ellipsoids of other planets can be very different than the Earth's ellipsoid).

It would probably not be reasonable to expect Proj4JS to contain the CRS definitions of all known planets of the universe. Even on Earth, it would probably not be reasonable to embedded the whole EPSG database with Proj4JS. Hence, the idea to plugin a mechanism which would query a server for fetching such information. Not everyone would want that, and those who want may want to query different servers (a server for astronomical bodies versus a server for accurate CRS on Earth, etc.). So that's why we thought about a plugin mechanism.

Above paragraphs talked about CRS definitions, but actually we are thinking about coordinate operations, not just CRS definitions. A CRS definitions could use the existing Proj4JS code, for example, for doing a Mercator projection with only different ellipsoid axis length. But sometime we need really different code. For example, Jupiter as a higher eccentricity than Earth, so the Taylor series expansion in the map projection formulas for Earth may not contain enough terms. We may want a different code with more terms. We are also thinking (but not sure if we will implement that soon) about triaxial ellipsoid. Even on Earth, an accurate coordinate operation may require a datum shift using something like a NADCON grid. There is different approach in using such grid. For example, NADCON applies translations in geographic coordinates, while French RGF93 applies translations in geocentric coordinates.

So in short, because we expect that Proj4JS would contain the most common operation methods but cannot contain all operation methods that may be desired, we are looking for a mechanism which would allow Proj4JS to query a server, and the server to return a JavaScript function doing the full coordinate operation (not only providing CRS definitions). Ideally, after a server has been registered, that mechanism would be transparent to a web application using Proj4JS.

**Comment by ahocevar** [comment]
Thanks for the context, @desruisseaux. I could imagine the API to be another signature of `proj4.defs()`, as simple as
```js
proj4.defs(code, forward, inverse);
```
with `forward()` and `inverse()` being functions:
```ts
([x: number, y: number]) => [x: number, y: number]
```
Then you could asynchronously get the `forward` and `inverse` functions from your server, and then call `proj4.defs()` with them.

How does that sound?

**Comment by jsorel** [comment]
The forward and inverse transforms are to/from latlong if I'm correct ?

**Comment by desruisseaux** [comment]
> The forward and inverse transforms are to/from latlong if I'm correct ?

I don't know if Proj4JS API has such expectation. But in general, it could be anything, not necessarily latitudes and longitudes.

**Comment by ahocevar** [comment]
`forward` is from `longlat`, inverse is to `longlat`.

**Comment by desruisseaux** [comment]
So what would like like an API from longlat to other longlat (datum shift), or from meters to other meters (change of map projection, e.g. when moving from one UTM zone to the next zone)?

**Comment by ahocevar** [comment]
Proj4js internally always goes from source projection to longlat to destination projection.

**Comment by desruisseaux** [comment]
> Proj4js internally always goes from source projection to longlat to destination projection.

I guess that what we would like to plugin is not the `forwardCustom` and `inverseCustom` functions then, but the whole work done by `proj4` in the particular case where `CUSTOM:1` and `CUSTOM:2` are handled by the server. The Proj4JS internal would be a fallback when the transform is not specified by a server.

**Comment by ahocevar** [comment]
That sounds like a very uncommon use case to me. Your [example above](https://github.com/proj4js/proj4js/issues/488#issuecomment-2476105066) also defines a transform with longlat as source.

**Comment by desruisseaux** [comment]
> That sounds like a very uncommon use case to me.

I'm not sure what you mean by unusual. Having something else than (latitude, longitude) as the source? It is not as common as map projections, but still common. The [approach that Proj4JS is using](https://github.com/proj4js/proj4js/blob/d653690fd6de5a0f664d4b560ac2a31cdee521ea/lib/transform.js#L29) is called _"early-binding"_ in EPSG terminology. It was the approach used by PROJ.4, before PROJ 6 moved to _"late-binding"_. Explaining what is "early-binding" versus "late-binding" would be a long story (some details [in this blog](https://www.geomatys.com/2017/09/20/proj-4-versus-apache-sis-an-accuracy-comparison/), in particular the "OSGB 1936 to ED50" example), but in short, "early-binding" is not recommended as it is impossible to achieve some degrees of accuracy with that approach. But we nevertheless recognize that "early-binding" is easier to implement, and acceptable when high accuracy is not a goal.

So in short, current Proj4JS design is fine for relatively "low" accuracy (a few meters). But if we want to give to users a chance to avoid the limitations of early-binding design without refactoring the whole Proj4JS project as a late-binding implementation (which would be a major work), a plugin mechanism for returning the whole transform would be desirable. The server would be a late-binding implementation. So it would be a way to provide late-binding benefits without rewriting large parts of Proj4JS.

**Comment by ahocevar** [comment]
May I ask which libraries except OpenLayers you are using with Proj4js? I'm still under the impression that when you have your exact transforms between the exact projections you're using, you don't need a library like Proj4js.

**Comment by jsorel** [comment]
> May I ask which libraries except OpenLayers you are using with Proj4js? I'm still under the impression that when you have your exact transforms between the exact projections you're using, you don't need a library like Proj4js.
> 

For now we only aim at OpenLayers. 
We are not web developers Martin and I, so I can't say which other tool/lib our web team will be using in the future.

**Comment by ahocevar** [comment]
OpenLayers allows exactly that kind of transforms between pairs of CRS. Proj4js (and PROJ as well) is designed for chaining transforms to and from longlat. So this feature request is beyond the scope of Proj4js. I think what you really want is a new library that implements an ISO 19111 transformations API, and could use Proj4js under the hood.

For OpenLayers, use the API as [suggested above](https://github.com/proj4js/proj4js/issues/488#issuecomment-2482351065).

**Comment by desruisseaux** [comment]
> Proj4js (and PROJ as well) is designed for chaining transforms to and from longlat.

Just for information, it was the case of Proj.4 but not anymore since PROJ 6. It was one of [GDAL barn](https://gdalbarn.com/) goals to resolve this Proj.4 limitation.

>  I think what you really want is a new library that implements an ISO 19111 transformations API, and could use Proj4js under the hood.

Not really. We have our own referencing engine, so we are looking for way to bypass Proj4JS but only when needed. This is an acknowledgement that Proj4JS is sufficient for common cases, while still trying to address more advanced cases. This is not an API issue, but really a question of capabilities.

While OpenLayer is the main current target, we thought that compatibility with Proj4JS API would be valuable. But I guess that a narrower focus on only OpenLayer can work at least for now.

Thanks
